### ISSUE TITLE
disconnection is sometimes unrecoverable without restarting the app

### ISSUE THREAD

**Issue by mrluc** [liveview-native/liveview-client-swiftui#41]
I was testing what happens when closing down the server started with `iex -S phx.server` and restarting, and ran into some crashes, and more frequently, unrecoverable error messages. (Nothing but ❤️ for this project, when it's connected it's already magical, totally understand that it's [v0.0.1-alpha software](https://dockyard.com/blog/2022/09/09/liveview-native-weekly-update-1), totally fine if answer is 'WONTFIX, wait for next version' or something. I'm documenting this in case it's helpful to someone. I'm curious if the details I mention here are known current limitations, or if those who have used this a bit more know something I don't -- like, is this only an issue in dev but not really in practice? maybe due to livereload paths being hit? etc -- be cool to look at the elixirconf ios chat app to see how it's handled there 😄).

**Observed in:** a working 'Getting Started' tutorial app, at the point of list > detail navigation, but before registering a custom component, [ie at the end of this tutorial]( https://liveviewnative.github.io/liveview-client-swiftui/tutorials/phoenixliveviewnative/03-navigation-and-hero)). Edit: in the Simulator, on an M1 Air.

**Problem:** killing the Elixir server, and then restarting it, often results in the app being 'stuck' in a broken state, at least in my experience. Restarting the mobile app fixes it.

This can happen in a variety of ways:
- 👍 If I restart the Elixir server very quickly, within 1-2 seconds, all is usually well despite the momentary error.
- 👎 👍  If I restart the Elixir server somewhat quickly, approx 3-5 seconds, the app gets backgrounded! Kind of funny. When I un-background it, it shows the error screen initially and then recovers.
- ❌ however, from the root view, if I kill the server and restart in about 10 seconds or more, I usually encounter this most-frequent kind of fail, where the UI disagrees with logging! The error continues to be displayed, but both the server and the XCode console show that reconnection succeeded, and log heartbeat messages, etc; screenie showing this case:

  <img width="1000" alt="image" src="https://user-images.githubusercontent.com/38204/189452186-458ee705-fd14-43b6-bd09-f92a48bf5872.png">
- 👍 likewise, if I disconnect the server while in a child view, see the error, and then restart the server after 20 seconds or so, the child view does not recover -- but when I navigate back to the parent view, it will sometimes look good, and continue to function normally (perhaps due to navigation triggering an explicit reconnection?)
- ❌ however, if I kill phx.server and wait a minute or two in the detail view, I sometimes experience a variety of harder fails, from the app exiting, to triggering this supposedly unreachable codepath in `NavStackEntryView.elementTree`
  <img width="1414" alt="image" src="https://user-images.githubusercontent.com/38204/189446954-038ff197-b243-437d-9a7b-b4dd1f8fcf43.png">
  
  - These warnings are usually associated in time with failed reconnection attempts:
    
    <img width="932" alt="image" src="https://user-images.githubusercontent.com/38204/189448474-00ef8d3e-6d9f-4700-adea-c6888e30eb25.png">

**Comment by bcardarella** [comment]
we should be handling long disconnects. Servers go down for more than a few seconds, people definitely background or close apps for more than a few seconds. So let's track this and it is definitely something we want to account for.

**Comment by mrluc** [comment]
Super cool. Amazing that you're taking this on w/just a few people 😄 par for the course this ecosystem though.

For future searchers, re: how to customize the 'error'/reconnection case -- it looks like that's doable by implementing [CustomRegistry.loadingView](https://liveviewnative.github.io/liveview-client-swiftui/documentation/phoenixliveviewnative/customregistry/loadingview(for:state:)-vg2v), haven't done that yet.

Off-topic -- I was thinking about a fun hack for offline, not for inclusion into LVN -- but it'd be fun to try to implement a 'fog of war' type of UI caching 😆 (fog of war, like in a real-time-strategy game), for the state of read-only views (this would have to be tagged/indicated somehow). That way, there could be an app-wide 'disconnected' indicator of some time, but you can still navigate around the cached/ghost views locally while you wait for it to come back, those you've interacted with in this session.

Hmm, now I'm curious about what [AirBnB's server-driven UI](https://medium.com/airbnb-engineering/a-deep-dive-into-airbnbs-server-driven-ui-system-842244c5f5) does when you go offline ...

**Comment by bcardarella** [comment]
A potential offline mode architecture may include Elixir Desktop. So we run the phoenix LV app locally on the device but the phoenix app acts a node in a cloud cluster. So you can still interact locally but then data syncs will occur when network is available. Just an idea for now though

**Comment by AZholtkevych** [comment]
Unblocked right now since https://github.com/liveview-native/phoenix-channels-client/issues/15 has been done
FYI @carson-katri @KronicDeth @Bajix 
@carson-katri you can proceed with that one

**Comment by carson-katri** [comment]
This is still blocked. `phoenix-channels-client` still needs to be integrated into this package.

**Comment by AZholtkevych** [comment]
@KronicDeth could you please help here

**Comment by KronicDeth** [comment]
@AZholtkevych what @carson-katri said is true.  `liveview-native` needs to use `phoenix-channels-client`.  That's https://github.com/liveview-native/liveview-native-core/issues/34 and https://github.com/liveview-native/liveview-native-core/issues/16, which is blocked by https://github.com/liveview-native/phoenix-channels-client/issues/45, which is what I'm currently working on.

**Comment by bcardarella** [comment]
This is likely resolved now that we have multiple app state phases to conditionally render
